File: January 2023/Day 11/Make array elements unique/make-array-elements-unique.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

#define ll long long int
class Solution {
  public:
    // Time Complexity :- O(n.log(n));
    // Space Complexity :- O(n);
    long long int minIncrements(vector<int> arr, int N) {
        // Code here
        sort(arr.begin(),arr.end());
        ll ans =0;
        set<ll> st;
        
        for(ll a : arr) {
            auto pos = st.find(a);
            if(st.find(a)!=st.end()) {
                ll val = *st.rbegin()+1;
                ans+=(val-a);
                a=val;
            }
            st.insert(a);
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin >> N;
        vector<int> arr(N);
        for (int i = 0; i < N; i++) {
            cin >> arr[i];
        }
        Solution obj;
        cout << obj.minIncrements(arr, N) << endl;
    }
}
// } Driver Code Ends

File: January 2023/Day 11/Make array elements unique/make-array-elements-unique.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int t =
            Integer.parseInt(br.readLine().trim()); // Inputting the testcases
        while (t-- > 0) {

            
            int i = 0;
            int N = Integer.parseInt(br.readLine().trim());
        
            int arr[] = new int[(int)(N)];
            
            String inputLine2[] = br.readLine().trim().split(" ");
            for ( i = 0; i < N; i++) {
                arr[i] = Integer.parseInt(inputLine2[i]);
            }
        
            Solution ob = new Solution();
            System.out.println(ob.minIncrements(arr,N));
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    public long minIncrements(int[] arr, int N) {
        // Code here
        Arrays.sort(arr);
        long ans =0;
        ArrayList<Long> list = new ArrayList<>();
        
        for(long a : arr) {
            if(isPresent(list,a)) {
                long val = list.get(list.size()-1)+1;
                ans+=(val-a);
                a = val;
            }
            list.add(a);
        }
        return ans;
    }
    
    public static boolean isPresent(ArrayList<Long> list,long val) {
        int lo=0,hi=list.size()-1;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(list.get(mid)<val) {
                lo=mid+1;
            } else if(list.get(mid)==val) {
                return true;
            } else {
                hi=mid-1;
            }
        }
        return false;
    }
}